feat: Add Twitter agent tutorial (#1122)
nearai/nearai: docs/agents/patterns/twitter/twitter_agent.md

#  Mention Twitter(X) Agent

NEAR AI allows anyone to easily create an [`X`](https://x.com) (Previously `Twitter`) agent that controls and account and answers to mentions.

![Demo](./images/demo.png)

> In this tutorial we will learn how the [twitter jokester agent](https://app.near.ai/agents/maguila.near/jokester/latest/source) works, and how you can change it to create your own X AI agent in less than five minute.

Reviewer: ```suggestion
In this tutorial, you'll learn how the [Twitter jokester agent](https://app.near.ai/agents/maguila.near/jokester/latest/source) works, and how you can change it to create your own X AI agent in less than five minutes.
```

---

## The X Agent

Let's start by exploring the code of the agent. The agent is a simple bot that replies to mentions with a joke, using the `tweepy` library to interact with the X API.

Try it out by tweeting a mention to the agent, and see how it replies with a joke or you can try it out by clicking [here](https://twitter.com/intent/tweet?text=Hey%20%40maguila_bot%2C%20tell%20me%20a%20joke%21)

```
Hey @maguila_bot, tell me a joke!
```

<hr class="subsection" />

### Invoking the Agent: Mentions

The agent works by listening to mentions of the account `@maguila_bot` and replying to them. The account that the agent listens to is configured in the `metadata.json` file, specifically in the section `x_mentions`.

```json title="metadata.json"
{
  "category": "agent",
  "description": "",
  "tags": [],
  "details": {
    "agent": {
      "welcome": {
        "description": "To use tweet a message and mention @maguila_bot.",
        "title": "Jokester"
      },
      "framework": "standard"
    },
    "triggers": {
      "events" : {
        "x_mentions": ["@maguila_bot"]
      }
    }
  },
  "show_entry": true,
  "name": "jokester",
  "version": "0.0.1"
}
```

!!! info

>     Technically speaking, the agent is not listening for mentions, but being executed when a specific event happens. In this case, when the account `@maguila_bot` is mentioned in `X` (denoted by the `event` `x_mentions`)

Reviewer: ```suggestion
    Technically speaking, the agent is not listening to mentions, but is executed when a specific event happens. In this case, when the account `@maguila_bot` is mentioned in `X` (denoted by the `event` `x_mentions`)
```

- tweet_id
- text

Since the `tweet` object contains all the information needed, we can process it through a model in order to generate a response.

``` python title="agent.py"
# Get the last message, which is the tweet
last_message = env.list_messages()[-1]

# Check if the last message is None
if last_message is None:
    env.add_reply("No message found")
    print("No message found")
    return

# Check if the last message is empty
if not last_message["content"]:
    env.add_reply("Message content was empty")
    print("Message content was empty")
    return

# Get the content of the message
event = json.loads(last_message["content"])

# Get the text of the tweet
input_user_text = event["text"]
# Get the id author of the tweet
input_tweet_id = event["tweet_id"]

# Generate a prompt
prompt = [
    {"role": "system", "content": PROMPT},
    {"role": "user", "content": input_user_text},
]

# Generate a response
joke = env.completion(messages=prompt, model=MODEL)

# Limit the length of the joke to 280 characters
if len(joke) > 280:
    joke = joke[:277] + "..."
```

<hr class="subsection" />

### Answering

The agent uses the `tweepy` library to send a reply to the tweet.

> First it creates a `tweepy` client, which is done in the `__init__` method of the agent. The client is created using the API keys stored in the environment variables.

Reviewer: ```suggestion
First, it creates a `tweepy` client, which is done in the `__init__` method of the agent. The client is created using the API keys stored in the environment variables.
```

``` python title="agent.py"
self.x_client = tweepy.Client(
          consumer_key=self.x_consumer_key,
          consumer_secret=self.x_consumer_secret,
          access_token=self.x_access_token,
          access_token_secret=self.x_access_token_secret
      )
```

Then, it uses the `send_tweet` method - which takes the `env`, `tweet`, and `tweet_id` - to send a reply to the tweet.

``` python title="agent.py"

def send_tweet(self, env, tweet, tweet_id):
    env.add_reply(f"Sending tweet: {tweet}")
    response = self.x_client.create_tweet(text=tweet, in_reply_to_tweet_id=tweet_id)
    print(f"Tweet published! ID: {response.data['id']}")

    # Send the tweet
    self.send_tweet(env, joke, input_tweet_id)
```

----

## Modifying the Agent

> If you want to create your own agent, you will need to start by forking [maguila](https://app.near.ai/agents/maguila.near/jokester/latest/source), and setting up the right Twitter API keys, so the agent can control the account you want.

Reviewer: ```suggestion
If you want to create your own agent, you will need to start by forking [Jokester](https://app.near.ai/agents/maguila.near/jokester/latest/source), and setting up the right Twitter API keys, so the agent can control the account you want.
```

<hr class="subsection" />

### Forking an Agent

> To fork the Twitter agent, enter [click here](https://app.near.ai/agents/maguila.near/jokester/latest/source) in your browser and click on the "Fork" button. This will create a copy of the agent in your NEAR AI registry.

Reviewer: ```suggestion
To fork the Twitter agent, [open this link](https://app.near.ai/agents/maguila.near/jokester/latest/source) in your browser and click the **`Fork`** button. This will create a copy of the agent in your NEAR AI registry.
```

![start](./images/02-fork-window.png)

<hr class="subsection" />

### Generate your API keys

To allow your agent to post to X you will need to get API keys from X and setup them in the agent. To generate your API keys, follow these steps:  

1. Go to the [Twitter Developer Portal](https://developer.twitter.com/en/portal/dashboard)
2. Click on the **"Projects & Apps"** tab.
3. Click on the **"Create App"** button.
4. Fill in the required fields, such as the app name and description.
5. Click on the **"Create"** button.
6. Once the app is created, go to the **"Keys and tokens"** tab.
7. You must create an **API Key** and **Secret**, and an **Access Token** and **Secret**.

!!! info "Free X Developer Account"
    Please know that free X developer accounts are limited on the amount of tweets they can produce.

!!! warning "Permissions"
    Remember to set the permissions for the Access Token and Secret to **"Read and Write"**.

    To change the permissions, go to the **Settings** tab, scroll down to the **User authentication settings** section, and select **"Read and Write"** under **App permissions**.

<hr class="subsection" />

### Setting up the Keys

> There are two ways to set your keys to environment variables. You can set them in web interface or in cli.

Reviewer: ```suggestion
There are two ways to set your keys to environment variables: through the web interface or through the CLI.
```

<hr class="subsection" />

#### Web interface

1. Go to the agent.
2. Click on the "Run" tab.
3. On the left side, click on the "Add" Button next to "Environment Variables".
4. Add the following variables:
    - `X_ACCESS_TOKEN`
    - `X_ACCESS_TOKEN_SECRET`
    - `X_CONSUMER_KEY`
    - `X_CONSUMER_SECRET`

Here is how it would look like:

![env](./images/env.png)

<hr class="subsection" />

#### CLI

> You can set your keys to environment variables using the cli.

Reviewer: ```suggestion
You can also set your keys to environment variables using the command-line interface (CLI).
```

> For example, if you want to set variable, you can run the following command in your terminal, remember to replace the values with your own keys:

Reviewer: ```suggestion
For example, if you want to set a variable, you can run the following command in your terminal (remember to replace the values with your own keys):
```

  -H "Content-Type: application/json" \
  -d '{
    "namespace": "maguila.near",
    "name": "jokester",
    "version": "0.0.1",
    "description": "consumer key for the X API",
    "key": "X_CONSUMER_KEY",
    "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
    "category": "agent"
  }'
```

```bash
curl -X POST "https://api.near.ai/v1/create_hub_secret" \
  -H "Authorization: Bearer {"account_id":"maguila.near","public_key":"ed25519:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","signature":"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","callback_url":"https://app.near.ai/sign-in/callback","message":"Welcome to NEAR AI Hub!","recipient":"ai.near","nonce":"00000000000000000000000000000000"}" \
  -H "Content-Type: application/json" \
  -d '{
    "namespace": "maguila.near",
    "name": "jokester",
    "version": "0.0.1",
    "description": "consumer secret for the X API",
    "key": "X_CONSUMER_SECRET",
    "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
    "category": "agent"
  }'
```

!!! info "Bearer Token"

>     You need generate a bearer token to use nearai api. you can read more about it [here](../../../api/guide.md).

Reviewer: ```suggestion
    You need generate a bearer token to use NEAR AI API. you can read more about it [here](../../../api/guide.md).
```

    
<hr class="subsection" />

### Modifying the Agent

> Let's change the agent into a charming 19th-century scholar who answers history questions with poetic poetic grace and Victorian elegance. You can also adjust the prompt and model accordingly, for example, like this:

Reviewer: ```suggestion
Let's change the agent into a charming 19th-century scholar who answers history questions with poetic grace and Victorian elegance. You can also adjust the prompt and model accordingly, for example, like this:
```
